### ISSUE TITLE
[Poison] Microsoft.DotNet.ILCompiler is consuming two PSB files

### ISSUE THREAD

**Issue by NikolaMilosavljevic** [dotnet/source-build#3181]
`Microsoft.DotNet.ILCompiler` is consuming PSB versions of `System.Collections.Immutable.dll` and `System.Reflection.Metadata.dll`.

From poison report:
```
Microsoft.DotNet.ILCompiler/tools/netstandard/System.Collections.Immutable.dll
Microsoft.DotNet.ILCompiler/tools/netstandard/System.Reflection.Metadata.dll
```
 
These files ship with SDK and `Microsoft.DotNet.ILCompiler` nupkg.

Here's why this is happening.

`Microsoft.DotNet.ILCompiler` project builds, in CoreCLR part of the build: https://github.com/dotnet/runtime/blob/main/src/coreclr/tools/aot/ILCompiler.Build.Tasks/ILCompiler.Build.Tasks.csproj.

It restores those two binaries via package reference to `System.Reflection.Metadata` (which depends on `System.Collections.Immutable`).

This causes the usage of PSB, as we don't yet have the new binaries or `Microsoft.DotNet.ILCompiler` package. Binaries will get built later in Libraries part of the build, and packaged during Installer part of the build.

The proposed solution is to modify runtime project build order to build these 2 libraries first and update package version in ILCompiler project.

Alternatively, it would technically be possible, but not desirable, to instruct the packaging project, to pick new files: https://github.com/dotnet/runtime/blob/d8e1ec686e8336b8e17288a0f1d3ebac1ce082d5/src/installer/pkg/projects/Microsoft.DotNet.ILCompiler/Microsoft.DotNet.ILCompiler.pkgproj#L27

**Comment by MichaelSimons** [comment]
@dotnet/crossgen-contrib, @ViktorHofer - Would appreciate your input on this and the possibility of re-ordering the runtime build order to address this issue.  TIA

**Comment by agocke** [comment]
As far as I know Native AOT isn't available in source build. Is there a separate reason why this is a problem?

**Comment by ViktorHofer** [comment]
Note that we have other examples of that happening today, i.e.: https://github.com/dotnet/runtime/blob/8e29ce70a41e1ffccc5f47d5d44ee36692098443/src/installer/managed/Microsoft.NET.HostModel/Microsoft.NET.HostModel.csproj#L24

Regarding `System.Reflection.Metadata`, according to [this comment](https://github.com/dotnet/runtime/blob/8e29ce70a41e1ffccc5f47d5d44ee36692098443/src/installer/managed/Microsoft.NET.HostModel/Microsoft.NET.HostModel.csproj#L22-L23) we intentionally don't use the live build version in some cases. What's the story for that?

**Comment by NikolaMilosavljevic** [comment]
> > It restores those two binaries via package reference to System.Reflection.Metadata (which depends on System.Collections.Immutable).
> 
> That's the main problem. That component should reference the live components in the repo instead of restoring prebuilts. I would imagine that replacing the PackageReference with one / two ProjectReference(s) should do the job.

System.Reflection.Metadata and System.Collections.Immutable projects are built later during runtime build. Some project ordering would be needed for this.

**Comment by ViktorHofer** [comment]
> System.Reflection.Metadata and System.Collections.Immutable projects are built later during runtime build. Some project ordering would be needed for this.

That's not true. You can just define a ProjectReference to System.Reflection.Metadata and build it at any time given that this targets netstandard2.0 which doesn't require reference assemblies to be built.

**Comment by NikolaMilosavljevic** [comment]
> > System.Reflection.Metadata and System.Collections.Immutable projects are built later during runtime build. Some project ordering would be needed for this.
> 
> That's not true. You can just define a ProjectReference to System.Reflection.Metadata and build it at any time given that this targets netstandard2.0 which doesn't require live reference assemblies being available.

OK, so the proposal is to remove this: https://github.com/dotnet/runtime/blob/8e29ce70a41e1ffccc5f47d5d44ee36692098443/src/coreclr/tools/aot/ILCompiler.Build.Tasks/ILCompiler.Build.Tasks.csproj#L23-L25

And, modify the following to reference project artifacts instead: https://github.com/dotnet/runtime/blob/8e29ce70a41e1ffccc5f47d5d44ee36692098443/src/coreclr/tools/aot/ILCompiler.Build.Tasks/ILCompiler.Build.Tasks.csproj#L28-L35

**Comment by ViktorHofer** [comment]
Yes, that's technically possible. You wouldn't fetch from the package cache but just make sure that the assemblies get copied to the output directory. Before talking about the concrete implementation though, we should first discuss this point:

> Regarding System.Reflection.Metadata, according to [this comment](https://github.com/dotnet/runtime/blob/8e29ce70a41e1ffccc5f47d5d44ee36692098443/src/installer/managed/Microsoft.NET.HostModel/Microsoft.NET.HostModel.csproj#L22-L23) we intentionally don't use the live build version in some cases. What's the source build story for that?

**Comment by jkotas** [comment]
> As far as I know Native AOT isn't available in source build

We do want to make it available. It is still partially building in source build. We just did not make it completely work for .NET 7.

**Comment by NikolaMilosavljevic** [comment]
> Note that we have other examples of that happening today, i.e.: https://github.com/dotnet/runtime/blob/8e29ce70a41e1ffccc5f47d5d44ee36692098443/src/installer/managed/Microsoft.NET.HostModel/Microsoft.NET.HostModel.csproj#L24
> 
> Regarding `System.Reflection.Metadata`, according to [this comment](https://github.com/dotnet/runtime/blob/8e29ce70a41e1ffccc5f47d5d44ee36692098443/src/installer/managed/Microsoft.NET.HostModel/Microsoft.NET.HostModel.csproj#L22-L23) we intentionally don't use the live build version in some cases. What's the source build story for that?

Source build uses source-built reference package, generated from the same version that runtime is using.

Should this package reference also be pinned to the same version as in the HostModel project (i.e.6.0.0, in main): https://github.com/dotnet/runtime/blob/8e29ce70a41e1ffccc5f47d5d44ee36692098443/src/coreclr/tools/aot/ILCompiler.Build.Tasks/ILCompiler.Build.Tasks.csproj#L23-L25
